lib/puppet/x509/cert_provider.rb

```diff
@@ -129,37 +138,45 @@ def save_private_key(name, key)
   #
   # @param name [String] The private key identity
   # @param required [Boolean] If true, raise if it is missing
+  # @param password [String, nil] If the private key is encrypted, decrypt
+  #   it using the password. If the key is encrypted, but a password is
+  #   not specified, then the key cannot be loaded.
   # @return (see #load_private_key_from_pem)
   # @raise (see #load_private_key_from_pem)
   # @raise [Puppet::Error] if the private key cannot be loaded
   # @api private
-  def load_private_key(name, required: false)
+  def load_private_key(name, required: false, password: nil)
     path = to_path(@privatekeydir, name)
     pem = load_pem(path)
     if !pem && required
       raise Puppet::Error, _("The private key is missing from '%{path}'") % { path: path }
     end
-    pem ? load_private_key_from_pem(pem) : nil
+    pem ? load_private_key_from_pem(pem, password: password) : nil
   rescue SystemCallError => e
     raise Puppet::Error.new(_("Failed to load private key for '%{name}'") % {name: name}, e)
   end
 
   # Load a PEM encoded private key.
   #
   # @param pem [String] PEM encoded private key
+  # @param password [String, nil] If the private key is encrypted, decrypt
+  #   it using the password. If the key is encrypted, but a password is
+  #   not specified, then the key cannot be loaded.
   # @return [OpenSSL::PKey::RSA, OpenSSL::PKey::EC] The private key
   # @raise [OpenSSL::PKey::PKeyError] The `pem` text does not contain a valid key
   # @api private
-  def load_private_key_from_pem(pem)
-    # set a non-nil passphrase to ensure openssl doesn't prompt
+  def load_private_key_from_pem(pem, password: nil)
+    # set a non-nil password to ensure openssl doesn't prompt
     # but ruby 2.4.0 & 2.4.1 require at least 4 bytes, see
     # https://github.com/ruby/ruby/commit/f012932218fd609f75f9268812df61fb26e2d0f1#diff-40e4270ec386990ac60d7ab5ff8045a4
```

Comment:
I know it wasn't added in this commit, but I believe that the GitHub link above shows the minimum number of bytes requirement being _removed_. Is this the right link to have here?

Reply:
It's the commit when the fix from ruby/openssl was merged back into ruby. Let me update that link so it's clearer.

Reply:
Is this a fix that went into ruby 2.4.2?

Reply:
Yep. The https://github.com/ruby/openssl repo contains the `openssl` gem which is periodically merged into ruby itself. The fix in `openssl` was merged into `ruby` in commit f012932218fd609f75f9268812df61fb26e2d0f1 and first released in ruby 2.4.2